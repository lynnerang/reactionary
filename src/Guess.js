import React, { Component } from 'react';
import './Guess.scss';

class Guess extends Component {
  constructor(props) {
    super(props);

    this.state = {
      errorTxt: '',
      errorIconClass: ''
    }
  }

  checkAnswer = (e) => {
    let value = e.target.type === 'button' ? e.target.id
    : e.target.closest('.guess-entry').querySelector('input').value;

    let correct = e.target.type === 'button' ? value === this.props.answer 
    : this.isSimilarEnough(value);

    if (correct) {
      this.showResponse('That is correct, nice job!', 'fa-check');
      this.props.removeCard();
    } else {
      this.showResponse('Not quite!  We\'ll try this one again later.', 'fa-times');
    }
    this.getNextTurn(e.target);
  }

  getNextTurn = () => {
    setTimeout(() => { 
      this.props.getRandomCard();
      const textInput = document.querySelector('#term-input');
      if (textInput) textInput.value = '';
      this.props.updateGuessCount(this.props.mode, 'add');
    }, 2000);
  }

  isSimilarEnough = (value) => {
    const check = require('string-similarity');
    const similarity = check.compareTwoStrings(value.toLowerCase(), this.props.answer.toLowerCase());
    return similarity >= .72;
  }

  showResponse = (text, icon) => {
    const response = document.querySelector('.guess-response');
    this.setState({errorTxt: text, errorIconClass: icon}, () => {
      response.classList.remove('invisible');
      setTimeout(() => response.classList.add('invisible'), 2000)
    })
  }

  onGuessKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.checkAnswer(e);
    }
  }

  render() {
    let iconClass = `fas ${this.state.errorIconClass}`;
    let guessLayout;

    this.props.mode === 'guess' ? guessLayout =  (
      <><label htmlFor='term-input'>Enter your guess:</label>
        <div className='guess-entry'>
          <input id='term-input' className='single-guess-input' autoComplete='off' onKeyPress={this.onGuessKeyPress}></input>
          <button type='button' className='single-guess-btn' onClick={this.checkAnswer}><i className='fas fa-arrow-right'></i></button>
        </div>
        <div className='guess-response correct-guess-response'>
          <i className={iconClass}></i><p>{this.state.errorTxt}</p>
        </div>
      </> )
    : guessLayout = (
      <><label>Choose a term:</label>
        <div className='guess-choices'>
          <button type='button' className='choice' id={this.props.choices[0]} onClick={this.checkAnswer}>{this.props.choices[0]}</button>
          <button type='button' className='choice' id={this.props.choices[1]} onClick={this.checkAnswer}>{this.props.choices[1]}</button>
          <button type='button' className='choice' id={this.props.choices[2]} onClick={this.checkAnswer}>{this.props.choices[2]}</button>
        </div>
        <div className='guess-response correct-guess-response'>
          <i className={iconClass}></i><p>{this.state.errorTxt}</p>
        </div>
      </> )

    return (
      <article className='guess-area'>
        {guessLayout}
      </article>
    )
  }
}

export default Guess;