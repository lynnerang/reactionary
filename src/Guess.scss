@import './variables.scss';

.guess-area {
  margin-top: 10px;
  margin-left: 40px;
  margin-bottom: 5px;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;

  label {
    padding-top: 25px;
    padding-bottom: 15px;
    font-size: 26px;
  }

  .guess-entry {
    display: flex;
  }

  .single-guess-input {
    padding-left: 10px;
    font-size: 18px;
    height: 40px;
    width: 225px;
    border-radius: 5px 0px 0px 5px;
    outline-color: $secondary-color;
    cursor: pointer;
  }

  .single-guess-btn {
    border-radius: 0px 5px 5px 0px;
    height: 40px;
    width: 40px;
    border: none;
    background: $secondary-color;
    cursor: pointer;
    outline-color: $secondary-color;

    &:hover,
    &:focus {
      background: $blue;
    }

    .fa-arrow-right {
      color: white;
      font-size: 20px;
    }
  }

  .guess-response {
    padding-top: 20px;
    padding-bottom: 40px;
    display: flex;
    align-items: center;

    p {
      padding-top: 3px;
    }

    .fas {
      margin-right: 10px;
    }
    
    .fa-check {
      font-size: 26px;
      color: rgb(65, 160, 65);
    }

    .fa-times {
      font-size: 30px;
      color: rgb(186, 52, 52);
    }
  }
}

.invisible {
  visibility: hidden;
}