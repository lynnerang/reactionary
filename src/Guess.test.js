import React from 'react';
import Guess from './Guess.js';
import { shallow } from 'enzyme';

const props = {
  showResponse: jest.fn(),
  updateGuessCount: jest.fn(),
  getRandomCard: jest.fn()
}

describe('Guess', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow (
      <Guess {...props}
      />
      )
  });

  it('Should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has two default state properties', () => {
    expect(wrapper.state()).toEqual({
      errorTxt: '',
      errorIconClass: ''
    });
  });

  it('can check a guess against an answer', () => {
    wrapper.instance().checkAnswer();
    expect(wrapper.checkAnswer).toHaveBeenCalledTimes(1);
  });

  it('can go to the next turn by updating the guess count and getting a new random card.', () => {
    wrapper.instance().getNextTurn('poop');

    expect(wrapper.getNextTurn).toHaveBeenCalledTimes(1);
    expect(props.updateGuessCount).toHaveBeenCalledTimes(1);
    expect(props.getRandomCard).toHaveBeenCalledTimes(1);
  });

  it('can calculate the level of similarity between the guess and answer.', () => {
    wrapper.instance().isSimilarEnough('poop');

    expect(wrapper.calcSimilarity).toHaveBeenCalledTimes(1);
  });

  it('can show a response in the dom depending on whether the guess was right.', () => {
    wrapper.instance().showResponse();

    expect(wrapper.showResponse).toHaveBeenCalledTimes(1);
  });

});