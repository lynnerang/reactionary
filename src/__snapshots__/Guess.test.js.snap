// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Guess Should match the snapshot 1`] = `undefined`;
